migration-challenge/migration-guide.md

```diff
@@ -0,0 +1,203 @@
+
+Purescript v0.15 is here and supports ESM with all of its advantages:
+
+* You can use ESM-only libraries
+* A cleaner and simpler way of writing FFI
+* Helping tools to do their job like [DCE](https://en.wikipedia.org/wiki/Dead_code_elimination) or [Code-Splitting](https://reactjs.org/docs/code-splitting.html#code-splitting)
+
+However, this has a couple of implications that will need you to migrate your code and tooling setup:
+
+* v0.15 drops support of CJS
+
+  With no real [reason](#a-little-bit-of-background) to keep supporting CJS, we have decided to drop CJS support alltogether. The maintenance burden would have been just too high to support CJS as well. This means you will need to update your FFI to ESM. More on this here:
+
+  [How can I update CJS to ESM?](#how-can-i-update-cjs-to-esm)
+
+
+* v0.15 drops support for Node.js versions < 12
+  
+  This is just the logical consequence of Node.js versions < 12 having reached EOL and not supporting ESM. More on this here:
+
+  [How can I use Purescript on Node.js?](#how-can-i-update-cjs-to-esm)
+
+* v0.15 drops `purs bundle` and relies on an external bundlers
+  
+  Yes, you heard right. The Purescript compiler no longer comes with a built-in `bundle` command. Updating `purs bundle` to ESM would have required considerable ammount of work, while at the same time being buggy, doing a poor job on bundle size and taking time from the compiler time to work on more useful things.
+
+  So v0.15 relies on an external bundler and that is good news because tools like `esbuild`, `webpack` or `parcel` do a much better job on bundling than `purs bundle`. You will see significantly improved bundle sizes with v0.15, like e.g. for [purescript-halogen template](https://github.com/purescript-halogen/purescript-halogen-template):
+
+  | Bundler         | Size |
+  |--------------|-----------|
+  | v0.14 with `purs bundle` | **259Kb** |
+  | v0.14 with `purs-bundle` and `esbuild` | **110Kb** |
+  | v0.15 with `esbuild` | **82Kb** |
+
+
+  More on this here:
+
+  [How can I bundle my library or application?](#how-can-i-bundle-my-library-or-application)
+
+## How can I update CJS to ESM?
+
+When you are writing JS FFI the most common situations where you will see changes are: 
+
+* Importing a module
+
+  In v0.14 you had to import a module using `require` 
+
+  ```javascript
+  const mymodule = require('mymodule')
+  ```
+
+  In v0.15 you need to use `import`
+
+  ```javascript
+  import * as M from 'mymodule'
+  // or import specific items from the module
+  import { main } from 'mymodule'
+  ```
+* Exporting variables and functions in your FFI
+
+  In v0.14 you had to use `exports`
+  ```javascript
+  exports.world = "world 🗺"
+  exports.greet = function() { return "hello" + world }
+  ```
+
+  In v0.15 you need to use `export` 
+  ```javascript
+  export const world = "world 🗺"
+  // or
+  export var world = "world 🗺"
+
+  export function greet() { return "hello" + world }
```

Comment:
Just FYI. There's currently a Discord conversation about whether to use explicitly-named functions or not and how that affects debugability:
```javascript
export var f = function(...) {...}
export const f = function(...) {...}
export const f = function f(...) {...}
export var f = (...) => {...}
export const f = (...) => {...}
export function f(...) {...}
```

Reply:
so what is the conclusion?

Reply:
Conclusion is it doesn't matter :tada: engines pick up the `name` with all these declarations (tested in FF, Chrome, node at least), and apparently it's part of the spec that they should do so now.

Reply:
In the possibility that `var` is ever deprecated sometime in the future because it gets replaced with `let`, I think using `const` should be our default. It'll save us the work of having to do the `var` -> `const` migration later.